File: Others/ccc16s3.java
import java.util.*;

public class ccc16s3 {

    static final int MAX = 100001;
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    static boolean[] pho = new boolean[MAX];
    static boolean[] valid = new boolean[MAX];
    static boolean[] vis = new boolean[MAX];
    static int[] dist = new int[MAX];
    static int diamater = 0;

    static void dfs(int u) {
        for(int v : graph.get(u)) {
            if(!vis[v] && valid[v]) {
                vis[v] = true;
                dist[v] = dist[u] + 1;
                dfs(v);
            }
        }
    }

    static boolean search(int u, int parent) {
        if(pho[u]) {
            valid[u] = true;
        }
        for(int v : graph.get(u)) {
            if(v != parent) {
                valid[u] |= search(v, u);
            }
        }
        return valid[u];
    }

    static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        for(int i = 0; i < m; i++) {
            int r = sc.nextInt();
            pho[r] = true;
        }

        for(int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for(int i = 0; i < n-1; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            graph.get(x).add(y);
            graph.get(y).add(x);
        }

        for(int i = 0; i < n; i++) {
            if(pho[i]) {
                search(i, -1);
                break;
            }
        }

        int count = n;
        Arrays.fill(vis, false);
        int src = 0;
        for(int i = 0; i < n; i++) {
            if(valid[i]) {
                src = i;
                break;
            }
        }
        dist[src] = 0;
        vis[src] = true;
        dfs(src);

        for(int i = src; i < n; i++) {
            if(valid[i] && dist[i] > dist[n]) {
                src = i;
            }
        }
        Arrays.fill(vis, false);
        dist[src] = 0;
        vis[src] = true;
        dfs(src);

        for(int i = 0; i < n; i++) {
            if(valid[i]) {
                diamater = Math.max(diamater, dist[i]);
            } else {
                count--;
            }
        }

        System.out.println((2 * (count - 1) - diamater));
    }
}

File: Others/ccc16s3.cpp
#include <bits/stdc++.h>
using namespace std;

#define f first
#define s second
#define pb push_back
#define FILL(a, b) memset(a, b, sizeof(a))

typedef long long int ll;
typedef pair<int, int> pii;
typedef pair<int, pii> piii;
typedef vector<int> vi;
typedef vector<pii> vii;

const int INF = 0x3f3f3f3f;
const ll INFL = 0x3f3f3f3f3f3f3f3fLL;
const int MOD = 1e9 + 7;
const int MAX = 1e5 + 5;

ll gcd(ll a, ll b){return b == 0 ? a : gcd(b, a % b);}
ll lcm(ll a, ll b){return a*b/gcd(a,b);}
ll fpow(ll b, ll exp, ll mod){if(exp == 0) return 1;ll t = fpow(b,exp/2,mod);if(exp&1) return t*t%mod*b%mod;return t*t%mod;}
ll divmod(ll i, ll j, ll mod){i%=mod,j%=mod;return i*fpow(j,mod-2,mod)%mod;}

int arr[MAX];
vector<int> adj[MAX];
bool pho[MAX];
bool vis[MAX];
bool onPath[MAX];
int dist[MAX];

// search all paths
void dfs(int u) {
    for (int v : adj[u]) {
        if (!vis[v] && onPath[v]) {
            vis[v] = true;
            dist[v] = dist[u] + 1;
            dfs(v);
        }
    }
}

// find if connected
bool unionFind(int u, int prev) {
    if (pho[u]) { onPath[u] = true; }
    for (int v : adj[u]) {
        if (v != prev) { onPath[u] |= unionFind(v, u); }
    }
    return onPath[u]; // compression
}

int main() {
    cin.sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    #ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
    #endif
    int N, M; cin >> N >> M;
    int n; 
    for (int i = 0; i < M; i++) {
        cin >> n;
        pho[n] = true;
    }

    for (int i = 0; i < N - 1; i++) {
        int x, y; cin >> x >> y;
        adj[x].push_back(y);
        adj[y].push_back(x);
    }

    for (int i = 0; i < N; i++) {
        if (pho[i]) { unionFind(i, -1); break; }
    }
    int cnt = N;
    FILL(vis, 0);
    n = 0;
    for (int i = 0; i < N; i++) {
        if (onPath[i]) { n = i; break; }
    }
    dist[n] = 0;
    vis[n] = true;
    dfs(n);
    for (int i = n; i < N; i++) {
        if (onPath[i] && dist[i] > dist[n]) { n = i; }
    }
    FILL(vis, 0);
    dist[n] = 0;
    vis[n] = true;
    dfs(n);

    int diam = 0;
    for (int i = 0; i < N; i++) {
        if (onPath[i]) { diam = max(diam, dist[i]); }
        else { cnt--; }
    }
    int ans = 2 * (cnt - 1) - diam;
    cout << ans << "\n";
    return 0;
}